### DISCUSSION TITLE
Are there any methods to correct outliers in RR or beat to beat intervals data?

### DISCUSSION THREAD

**Discussion by apar945** [neuropsychology/NeuroKit#828] [Q&A]
Hello 

Are there any function for artifact correction, ectopic beats removal and outliers removal (RR intervals less than 400ms or above 2000ms)?

This feature is available in Kubios software but it is a blackbox and hence we have no idea how it's done.

Another library is https://github.com/Aura-healthcare/hrv-analysis
It has ectopic beats removal and outliers removal but not artifact correction. 

Regards
Ankit

**Comment by danibene** [comment]
Hi @apar945

I would suggest looking at `signal_fixpeaks()` and `find_outliers()`

https://github.com/neuropsychology/NeuroKit/blob/master/neurokit2/signal/signal_fixpeaks.py#L222-L225

https://github.com/neuropsychology/NeuroKit/blob/master/neurokit2/misc/find_outliers.py

**Reply by LoraGee** [reply]
Hey @apar945  ,
I also have collected RR intervals and want to use neurokit2 for my PhD analysis. Facing similar problems. 
Are you interested in a catch-up?

Best
Lora

**Reply by chandlermassey4** [reply]
Hi @apar945 and @LoraGee,

I am was wondering if either of you had compared the HRV values you get from the neurokit's *hrv* function to the values you get using Kubios. When I compare my own data I get almost exact values for HRV_MeanNN, and RMSSD, but other values seem to be random.

**Reply by mdd8317** [reply]
> Hey @apar945 , I also have collected RR intervals and want to use neurokit2 for my PhD analysis. Facing similar problems. Are you interested in a catch-up?
> 
> Best Lora

@LoraGee, I did not use Neurokit2. I referred to various research publications and created my own custom outlier removal algorithm to clean the data. 

https://github.com/Aura-healthcare/hrv-analysis
This library has documented some methods for outlier detection and Karlsson method is effective in my case. I compared PPG sensors against a clinical ECG system and my implementation is very specific to my use case. 

Here are some helpful links to correcting or rather interpolating data once you have identified outliers. 
https://hal.science/hal-03501191/document
https://www.frontiersin.org/journals/physiology/articles/10.3389/fphys.2012.00148/full
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6679245/

Hope this helps.

**Reply by mdd8317** [reply]
> Hi @apar945 and @LoraGee,
> 
> I am was wondering if either of you had compared the HRV values you get from the neurokit's _hrv_ function to the values you get using Kubios. When I compare my own data I get almost exact values for HRV_MeanNN, and RMSSD, but other values seem to be random.

@chandlermassey4, Kubios is a black-box system so it won't be clear what outlier correction is applied before HRV values are calculated. Before you compare the systems, go through the documentation to understand how the data is manipulated. The mathematical formula for time and frequency-based HRV calculations are standard. It is the part prior to calculation that involves detecting beats per minute, ectopic beat outliers that would differ.

**Reply by chandlermassey4** [reply]
Hi @mdd8317,

Since you are pursuing a phd in this area, do you know if there are any data set online that include a raw ECG signal and verified HRV metric for that signal? 

I ask this because I have extracted the RR intervals that Kubios generates to an output file and am getting different values for some hrv features mainly frequency and non linear domain when I generate features using Neurokit. I am unsure whether Kubios has already identified and removed ectopic beats, but when I am examine the intervals it does not appear any outliers exist which adds to my confusion of why the metric are so different.

**Reply by mdd8317** [reply]
@chandlermassey4 

I did not seek one. However, try the Physionet database
https://physionet.org/content/?topic=HRV+database

Compare Neurokit output to https://github.com/Aura-healthcare/hrv-analysis
Here you are able to compare code as well to see if there is anything different.
